tests/virt_autotest/xen_guest_irqbalance.pm

```diff
@@ -0,0 +1,192 @@
+
+        my $nproc = script_output("ssh root\@$guest 'nproc'");
+
+        # Bind all NIC IRQs to cpu0 in guest
+        my @nic_irqs_id = split('\n', script_output("ssh root\@$guest \"grep vif /proc/interrupts | cut -d ':' -f1 | sed 's/^ *//'\""));
+        foreach my $irq_id (@nic_irqs_id) {
+            assert_script_run("ssh root\@$guest \"echo 1 > /proc/irq/$irq_id/smp_affinity\"");
+        }
+        my @affinities_with_binding_cpu0 = get_irq_affinities_from_guest($guest, @nic_irqs_id);
+        foreach (@affinities_with_binding_cpu0) {
+            die "Failed to bind NIC IRQs to CPU0" if $_ ne 1;
+        }
+
+        # Start the irqbalance.service again
+        assert_script_run("ssh root\@$guest \"systemctl start irqbalance\"");
+
+        # Check if the SMP affinities are distributed among CPUs
+        my @affinities_with_irqbalance = get_irq_affinities_from_guest($guest, @nic_irqs_id);
+        # for SLE15SP3 and prior, IRQs are distributed on one cpu core ramdomly.
+        # So it is correct if some of them may happen to run on cpu0
+        # but it would not be correct if all IRQs were bound to cpu0.
+        if (guest_is_sle($guest, '<=15-sp3')) {
+            my $multiply = 1;
+            $multiply *= $_ foreach (@affinities_with_irqbalance);
+	    print "julie: \$multiply=$multiply of affinities for $guest after irqbalance.\n";
+            die "SMP affinities were all bound to CPU0 with irqbalance.service enabled" if $multiply == 1;
+        }
+        # for SLE15SP4 (or newer release), IRQs are specified on all CPU cores
+        # fg. affinity value is '0xf' for machine with 4 CPU cores
+        else {
+            foreach (@affinities_with_irqbalance) {
+                die "It is not correct that one of the NIC IRQs was bound to cpu0 with irqbalance enabled." if $_ eq 1;
+            }
+        }
+
+        # Get the network interrupts distribution over CPUs in guest prior to network loads
+        my @initial_total_irqs_on_cpu = get_nic_irqs_distribution_from_guest($guest, $nproc);
+
+        # Generate network load in guest
+        assert_script_run("ssh root\@$guest \"ping -f -c 500 download.opensuse.org\"");
+
+        # re-caculate the network interrupts distribution
+        # the increased NIC IRQs should be distributed in CPU cores in balance
+        my @total_irqs_on_cpu_after_network_download = get_nic_irqs_distribution_from_guest($guest, $nproc);
+        my @increased_irqs_on_cpu;
+        for (my $cpu_id = 0; $cpu_id < $nproc; $cpu_id++) {
+            $increased_irqs_on_cpu[$cpu_id] = $total_irqs_on_cpu_after_network_download[$cpu_id] - $initial_total_irqs_on_cpu[$cpu_id];
+            die("The value of network interrupts from download for CPU" . $cpu_id . " is 0") if $increased_irqs_on_cpu[$cpu_id] == 0;
+        }
+        record_info("IRQs of downloads distribution", "@increased_irqs_on_cpu");
+    }
+
+    restore_original_guests();
+
+}
```

Comment:
I think your workflow can be organized better, for example, do everything in the order of set-do-check and group relevant code into respective functions.

Reply:
Have wrapped some functions per comments, such as save_original_guests() and restore_original_guests() to make the scenario clearer. The remained blocks are smaller and should not be called twice. Thanks.